### PR TITLE
Use serialized plan hash for fragment result caching

### PR DESCRIPTION
**Pull request by shixuan-fan**
Test plan - Enhanced existing unit test. Verifier run for internal queries passed.

```
== RELEASE NOTES ==

General Changes
* Reduce memory usage for fragment result caching.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java [hunk 1]
```diff
@@ -85,4 +106,9 @@ public CanonicalPlanFragment getCanonicalPlanFragment()
     {
         return canonicalPlanFragment;
     }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java:108]: remove this method?

File: presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java [hunk 2]
```diff
@@ -24,34 +25,53 @@
...
+            Session session,
+            ObjectMapper objectMapper)
     {
         if (!SystemSessionProperties.isFragmentResultCachingEnabled(session) || !isEligibleForFragmentResultCaching(root)) {
             return Optional.empty();
         }
 
         Optional<CanonicalPlanFragment> canonicalPlanFragment = generateCanonicalPlan(root, partitioningScheme);
-        return canonicalPlanFragment.map(fragment -> new FragmentResultCacheContext(fragmentResultCacheManager, fragment));
+        if (canonicalPlanFragment.isPresent()) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java:62]:
```java
if (!canonicalPlanFragment.isPresent()) {
    return Optional.empty();
}

try {
...

**Review by highker — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java (hunk 1 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/execution/FragmentResultCacheContext.java:108]: For the scope of this PR, we could remove `canonicalPlanFragment` and its getter method in this class. We probably would introduce them back when trying to support partition-stats based predicate stripping.

**Comment by gravitys169** [comment]
@shixuan-fan ,is there a plan of expland the fragment result cache to other stage which use remote split? so it may support more kinds of aggregation sql?

**Comment by shixuan-fan** [comment]
@gravitys169 : As chatted in Slack DM, no we currently don't plan to support multiple stages for fragment result caching.

**Comment by gravitys169** [comment]
got it,thanks!

**Pull request merged**